public/app/features/plugins/sandbox/sandbox_plugin_loader.ts

```diff
@@ -0,0 +1,180 @@
+    return generalDistortionMap.get(v) ?? v;
+  }
+
+  return new Promise(async (resolve, reject) => {
+    const isDevMode = config.buildInfo.env === 'development';
+
+    // each plugin has its own sandbox
+    const sandboxEnvironment = createVirtualEnvironment(window, {
+      // distortions are interceptors to modify the behavior of objects when
+      // the code inside the sandbox tries to access them
+      distortionCallback,
+      // endowments are custom variables we make available to plugins in their window object
+      endowments: Object.getOwnPropertyDescriptors({
+        // Plugins builds use the AMD module system. Their code consists
+        // of a single function call to `define()` that internally contains all the plugin code.
+        // This is that `define` function the plugin will call.
+        // More info about how AMD works https://github.com/amdjs/amdjs-api/blob/master/AMD.md
+        // Plugins code normally use the "anonymous module" signature: define(depencies, factoryFunction)
+        define(
+          idOrDependencies: string | string[],
+          maybeDependencies: string[] | PluginFactoryFunction,
+          maybeFactory?: PluginFactoryFunction
+        ): void {
+          let dependencies: string[];
+          let factory: PluginFactoryFunction;
+          if (Array.isArray(idOrDependencies)) {
+            dependencies = idOrDependencies;
+            factory = maybeDependencies as PluginFactoryFunction;
+          } else {
+            dependencies = maybeDependencies as string[];
+            factory = maybeFactory!;
+          }
+
+          try {
+            const resolvedDeps = resolvePluginDependencies(dependencies);
+            // execute the plugin's code
+            const pluginExports: { plugin: GrafanaPlugin } = factory.apply(null, resolvedDeps);
+            // only after the plugin has been executed
+            // we can return the plugin exports.
+            // This is what grafana effectively gets.
+            resolve(pluginExports);
+          } catch (e) {
+            reject(new Error(`Could not execute plugin ${meta.id}: ` + e));
+          }
+        },
+      }),
+      // This improves the error message output for plugins
+      // because errors thrown inside of the sandbox have a stack
+      // trace that is difficult to read due to all the sandboxing
+      // layers.
+      instrumentation: {
+        // near-membrane concept of "activity" is something that happens inside
+        // the plugin instrumentation
+        startActivity() {
+          return {
+            stop: () => {},
+            error: getActivityErrorHandler(meta.id),
+          };
+        },
+        log: () => {},
+        error: () => {},
+      },
+    });
+
+    // fetch and evalute the plugin code inside the sandbox
+    try {
+      let pluginCode = await getPluginCode(meta.module);
+      if (isDevMode) {
+        pluginCode = patchPluginSourceMap(meta, pluginCode);
```

Comment:
Currently we are serving the source maps in production as well - wouldn't it make sense to patch the plugin code both in dev and prod?

Reply:
Good point!